tests/suite/test_app_protect_grpc.py

```diff
@@ -0,0 +1,168 @@
+    create_example_app(kube_apis, "grpc", test_namespace)
+    wait_until_all_pods_are_ready(kube_apis.v1, test_namespace)
+
+    print("------------------------- Deploy Secret -----------------------------")
+    src_sec_yaml = f"{TEST_DATA}/appprotect/appprotect-secret.yaml"
+    create_items_from_yaml(kube_apis, src_sec_yaml, test_namespace)
+
+    print("------------------------- Deploy logconf -----------------------------")
+    src_log_yaml = f"{TEST_DATA}/appprotect/logconf.yaml"
+    log_name = create_ap_logconf_from_yaml(kube_apis.custom_objects, src_log_yaml, test_namespace)
+
+    print(f"------------------------- Deploy appolicy: {policy} ---------------------------")
+    src_pol_yaml = f"{TEST_DATA}/appprotect/grpc/{policy}.yaml"
+    pol_name = create_ap_policy_from_yaml(kube_apis.custom_objects, src_pol_yaml, test_namespace)
+
+    print("------------------------- Deploy Syslog -----------------------------")
+    src_syslog_yaml = f"{TEST_DATA}/appprotect/syslog.yaml"
+    create_items_from_yaml(kube_apis, src_syslog_yaml, test_namespace)
+    wait_before_test(40)
+    syslog_ep = (
+            kube_apis.v1.read_namespaced_endpoints("syslog-svc", test_namespace)
+            .subsets[0]
+            .addresses[0]
+            .ip
+        )
+    
+    print("------------------------- Deploy ingress -----------------------------")
+    ingress_host = {}
+    src_ing_yaml = f"{TEST_DATA}/appprotect/grpc/ingress.yaml"
+    create_ingress_with_ap_annotations(kube_apis, src_ing_yaml, test_namespace, policy, "True", "True", f"{syslog_ep}:514")
+    ingress_host = get_first_ingress_host_from_yaml(src_ing_yaml)
+    wait_before_test(40)
+
+    def fin():
+        print("Clean up:")
+        delete_items_from_yaml(kube_apis, src_syslog_yaml, test_namespace)
+        delete_items_from_yaml(kube_apis, src_ing_yaml, test_namespace)
+        delete_ap_policy(kube_apis.custom_objects, pol_name, test_namespace)
+        delete_ap_logconf(kube_apis.custom_objects, log_name, test_namespace)
+        delete_common_app(kube_apis, "grpc", test_namespace)
+        delete_items_from_yaml(kube_apis, src_sec_yaml, test_namespace)
+        replace_configmap_from_yaml(kube_apis.v1,
+                        ingress_controller_prerequisites.config_map['metadata']['name'],
+                        ingress_controller_prerequisites.namespace,
+                        f"{DEPLOYMENTS}/common/nginx-config.yaml")
+
+    request.addfinalizer(fin)
+
+    return BackendSetup(ingress_host, ingress_controller_endpoint.port_ssl)
+
+
+@pytest.mark.skip_for_nginx_oss
+@pytest.mark.appprotect
+@pytest.mark.smoke
+@pytest.mark.parametrize(
+    "crd_ingress_controller_with_ap",
+    [{"extra_args": [f"-enable-custom-resources", f"-enable-app-protect"]}],
+    indirect=["crd_ingress_controller_with_ap"],
+)
+class TestAppProtect:
+    @pytest.mark.parametrize("backend_setup", [{"policy": "grpc-block-sayhello"}], indirect=True)
```

Comment:
I wonder if the approach of instantiating the `backend_setup` two times is optimal. Since we have multiple 40s sleeps there, I think calling `backend_setup` once instead of twice will save a reasonable amount of time. 

I wonder if it makes sense just to have a singe test that (1) tests the block sayhello policy (2) updates the policy to block saygoodbye (3) tests the block saygoodbye policy. Would it save time? Would the saving worth it?

Reply:
@pleshakov I tried this approach of having one test before but imo It'd be better to have both of them independent of each other since we'd still need to wait after applying policy and ingress. There will be however only one wait needed for syslog pod but having separate syslog deployments makes the log more readable as its just a massive string not a structured log format like json. Also see: https://github.com/nginxinc/kubernetes-ingress/pull/1603#discussion_r635115886

Reply:
I'm assuming the 40s that is catching us out is app protect ?